### PR TITLE
Refactor istio agent

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Mainly does:

1. split the agent options to ./options
1. simplify envoy agent as no restart is needed.

### CODE REVIEW

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -79,63 +72,13 @@ var (
 	podNameVar           = env.RegisterStringVar("POD_NAME", "", "")
 	podNamespaceVar      = env.RegisterStringVar("POD_NAMESPACE", "", "")
 	kubeAppProberNameVar = env.RegisterStringVar(status.KubeAppProberEnvName, "", "")
-	serviceAccountVar    = env.RegisterStringVar("SERVICE_ACCOUNT", "", "Name of service account")
```
Comment (ramaraochavali) [inline pilot/cmd/pilot-agent/main.go:82]: instanceIPvar and others have been moved to options.go? Seems redundant here?

**Review by ramaraochavali — COMMENTED** [review]
LGTM. One minor comment

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/main.go:82]: Good catch, i am thinking further moving all the envs and even flags out and leaves the main file clear.

**Comment by mandarjog** [comment]
@kyessenov PTAL

**Comment by mandarjog** [comment]
@kyessenov this is removing the ability to do hot restarts completely. I think it is fine, but PTAL.

**Comment by kyessenov** [comment]
@mandarjog Hot restart has been removed long time ago. I think this is just a clean up of dead code.

File: pilot/cmd/pilot-agent/options/agent.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// Copyright Istio Authors
```
Comment (costinm) [inline pilot/cmd/pilot-agent/options/agent.go:1]: This should probably not be under pilot/cmd - maybe in pkg/isito-agent ?

**Review by costinm — COMMENTED** [review]
I like many of the refactorings - but please don't mix removal of functionality with cleanup/moving things around. 

Also removing features like hot restart should be discussed / approved by Net or Env WG - I am not entirely sure it is no longer needed. I know we're not using it now - because we don't support hot restart, but we may still need to do it.

**Review by costinm — COMMENTED** [review]
On further thought - I am personally ok with removing the hot restart implementation, it is over complicated. If/when we support hot restart - we can add a simplified version. 

+1 on the removal - see my other comment on location, not sure how we ended up with packages under /cmd

**Comment by howardjohn** [comment]
I agree with costin - I know we don't use hot restart now, and maybe YAGNI it applies, but we have been very close to using it for a long time. I have seen a number of prototypes over the past year or so utilizing hot restart, so it may be a bit premature to remove it entirely.

The rest looks good though

File: pkg/istio-agent/agent.go [hunk 3]
```diff
@@ -91,11 +91,11 @@ type Agent struct {
 	localDNSServer *dns.LocalDNSServer
 }
 
-// AgentConfig contains additional config for the agent, not included in ProxyConfig.
+// AgentOptions contains additional config for the agent, not included in ProxyConfig.
 // Most are from env variables ( still experimental ) or for testing only.
 // Eventually most non-test settings should graduate to ProxyConfig
 // Please don't add 100 parameters to the NewAgent function (or any other)!
-type AgentConfig struct {
+type AgentOptions struct {
```
Comment (costinm) [inline pkg/istio-agent/agent.go:98]: Should we just move them to ProxyConfig and get rid of the extra options ?

**Comment by costinm** [comment]
Also since we are moving files around - it may be a good idea at some point ( not this PR) to move all agent related files to a separate dir ( from to agent/cmd and agent/pkg ) - possibly with a separate go.mod to keep the bloat out.

File: pkg/istio-agent/agent.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pkg/istio-agent/agent.go:98]: I think this was to make it just own only the necessary fields

File: pilot/cmd/pilot-agent/options/agent.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/options/agent.go:1]: Personally I donot like to move options under pkg. I'd like to move it and also the main under istio-agent/xxx .

**Comment by hzxuzhonghu** [comment]
> possibly with a separate go.mod to keep the bloat out.

I haven't investigated how to do this within one repo.

**Comment by howardjohn** [comment]
I think you will struggle to isolate the dependencies if istio-agent in any
meaningful way. We already do not import k8s client libraries but we still
import almost all of istiod and envoy.

I guess if our goal is to import the agent without k8s (but still with 40mb
of other dependencies) it would help - but not sure it's worth the
complexity of nested projects?

**Comment by hzxuzhonghu** [comment]
@howardjohn What's the tool you use to analyze the dependency? Can we let this go and i want to further simplify the pilot-agent,  the memory cost of it is too much.

istio-p+  787494  0.0  0.2 742352 43776 ?        Ssl  Mar16   0:53 /usr/local/bin/pilot-agent proxy sidecar --domain default.svc.cluster.local --serviceCluster httpbin.default --proxyLogLevel=warning --proxyComponentLogLevel=misc:error --log_output_level=cache:debug --concurrency 2

**Comment by howardjohn** [comment]
I use `goda`. What dependency do you think we can drop?

**Comment by hzxuzhonghu** [comment]
From the dependency tree `k8s.io/client-go` is imported, which should de unnecessary. This is because of sds server make use of `istio.io/istio/pilot/pkg/xds` i guess.

**Comment by howardjohn** [comment]
A while back I dropped the import on k8s.io/api which was the biggest -
about 30mb to the final binary. IIRC, when I was looking into this
previously the remaining k8s imports (api machinery) are fairly small and
pretty hard to remove so I didn't pursue it any further.

**Comment by hzxuzhonghu** [comment]
From goda tree k8s.io/api is still imported,

**Comment by howardjohn** [comment]
oh you need to add -tags=agent, that is how we build it to drop those
imports

**Comment by hzxuzhonghu** [comment]
k8s.io/client-go --> import k8s.io/api. Still some dependencies.

**Comment by howardjohn** [comment]
```
$ goda tree 'agent=1(reach(./pilot/cmd/pilot-agent, k8s.io/api/...:root))'
$ goda tree 'agent=1(reach(./pilot/cmd/pilot-agent, k8s.io/client-go/...:root))'
  ├ istio.io/istio/pilot/cmd/pilot-agent
    └ istio.io/istio/pkg/istio-agent
      └ istio.io/istio/security/pkg/nodeagent/sds
        └ istio.io/istio/pilot/pkg/xds
          └ istio.io/istio/pilot/pkg/controller/workloadentry
            ├ k8s.io/client-go/tools/cache
              ├ k8s.io/client-go/rest
                ├ k8s.io/client-go/pkg/apis/clientauthentication
                ├ k8s.io/client-go/pkg/version
                ├ k8s.io/client-go/plugin/pkg/client/auth/exec
                  ├ k8s.io/client-go/pkg/apis/clientauthentication ~
                  ├ k8s.io/client-go/pkg/apis/clientauthentication/v1alpha1
                    └ k8s.io/client-go/pkg/apis/clientauthentication ~
                  ├ k8s.io/client-go/pkg/apis/clientauthentication/v1beta1
                    └ k8s.io/client-go/pkg/apis/clientauthentication ~
                  ├ k8s.io/client-go/tools/clientcmd/api
                  ├ k8s.io/client-go/tools/metrics
                  ├ k8s.io/client-go/transport
                    ├ k8s.io/client-go/util/connrotation
                    └ k8s.io/client-go/util/workqueue
                  └ k8s.io/client-go/util/connrotation ~
                ├ k8s.io/client-go/rest/watch
                ├ k8s.io/client-go/tools/clientcmd/api ~
                ├ k8s.io/client-go/tools/metrics ~
                ├ k8s.io/client-go/transport ~
                ├ k8s.io/client-go/util/cert
                  └ k8s.io/client-go/util/keyutil
                └ k8s.io/client-go/util/flowcontrol
              └ k8s.io/client-go/tools/pager
            └ k8s.io/client-go/util/workqueue ~

```

**Comment by howardjohn** [comment]
and cutting those only drops 400kb, so not much at all

**Comment by howardjohn** [comment]
`goweight --tags=agent ./pilot/cmd/pilot-agent` is another good one

**Comment by costinm** [comment]
We don't have to do everything in this PR - we can remove more deps and cleanup later.

My only remaining concern is the location of the packages under pilot/cmd/... - but I'm ok if we move them to /pkg in a separate PR as well. For the rest - +1 on this PR, and please keep doing cleanups (in smaller PRs if possible) !

**Comment by hzxuzhonghu** [comment]
Yes, will continue this work in follow ups.

**Comment by hzxuzhonghu** [comment]
>and cutting those only drops 400kb, so not much at all

@howardjohn  How do you get 400KB? With `goweight`? Basically, i donot mind the binary size much, I want to reduce the memory cost by pilot-agent.

**Comment by howardjohn** [comment]
I removed the dependencies and compared the built binary size with `stat`. Isn't the memory usage directly proportional to the binary size? We are not using the k8s libraries at runtime even if they are in the library

**Comment by mandarjog** [comment]
The memory size is a combination of binary size (shared) and unshareable memory. The effect of large binary size is reduced in k8s due to sharing. The unshared memory is duplicated. Kubectl top on a node with realistic number of pods gives the correct numbers.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**